{% extends "base.html" %}
{% block title %} References {% endblock %}
{% block meta_description %} References {% endblock %}
{% block content %}

<div id="references" class="width-960">
  <div id="hero" class="dark short">
    <div id="hero-content" class="wide">
      <h1 id="logo-title">
        Hail-Powered Science
      </h1>
      <div class="logo-subtitle">
        An incomplete list of scientific work enabled by Hail.
      </div>
    </div>
  </div>
  <div class="about">
    <div class="content">
      <p>If you use Hail for published work, please cite the software. You can get a citation for the version of Hail you installed by executing:</p>
      <pre><code class="language-python">import hail as hl
print(hl.citation())</code></pre>
      <p>Or you could include the following line in your bibliography:</p>
      <pre><code>Hail Team. Hail 0.2. https://github.com/hail-is/hail</code></pre>
      <p>Otherwise, we welcome you to add additional examples by <a href="https://github.com/hail-is/hail/edit/main/website/website/pages/references.html">editing this page directly</a>, after which we will review the pull request to confirm the addition is valid. Please adhere to the existing formatting conventions.</p>
      <p><em>Last updated on February 22, 2024</em></p>
      <h2 id="section">2024</h2>
      <ul>
        <li>
          <p>
	    Kwak, S.H., Srinivasan, S., Chen, L. et al. Genetic architecture and biology of
	    youth-onset type 2 diabetes. Nat Metab 6, 226–237
	    (2024). <a href="https://doi.org/10.1038/s42255-023-00970-0">https://doi.org/10.1038/s42255-023-00970-0</a>
	    <a href="https://www.nature.com/articles/s42255-023-00970-0">https://www.nature.com/articles/s42255-023-00970-0</a>
          </p>
        </li>
        <li>
          <p>
	    Zhao, S., Crouse, W., Qian, S. et al. Adjusting for genetic confounders in
	    transcriptome-wide association studies improves discovery of risk genes of complex
	    traits. Nat Genet 56, 336–347
	    (2024). <a href="https://doi.org/10.1038/s41588-023-01648-9">https://doi.org/10.1038/s41588-023-01648-9</a>
	    <a href="https://www.nature.com/articles/s41588-023-01648-9">https://www.nature.com/articles/s41588-023-01648-9</a>
          </p>
        </li>
      </ul>
      <h2 id="section">2023</h2>
      <ul>
        <li>
          <p>
	    Lee, S., Kim, J. & Ohn, J.H. Exploring quantitative traits-associated copy number
	    deletions through reanalysis of UK10K consortium whole genome sequencing cohorts. BMC
	    Genomics 24, 787 (2023). <a href="https://doi.org/10.1186/s12864-023-09903-3">https://doi.org/10.1186/s12864-023-09903-3</a> <a href="https://link.springer.com/article/10.1186/s12864-023-09903-3">https://link.springer.com/article/10.1186/s12864-023-09903-3</a>
          </p>
        </li>
        <li>
          <p>
	    Langlieb, J., Sachdev, N.S., Balderrama, K.S. et al. The molecular cytoarchitecture of
	    the adult mouse brain. Nature 624, 333–342
	    (2023). <a href="https://doi.org/10.1038/s41586-023-06818-7">https://doi.org/10.1038/s41586-023-06818-7</a>
	    <a href="https://www.nature.com/articles/s41586-023-06818-7">https://www.nature.com/articles/s41586-023-06818-7</a>
          </p>
        </li>
        <li>
          <p>
	    Leońska-Duniec, A., Borczyk, M., Korostyński, M. et al. Genetic variants in myostatin
	    and its receptors promote elite athlete status. BMC Genomics 24, 761
	    (2023). <a href="https://doi.org/10.1186/s12864-023-09869-2">https://doi.org/10.1186/s12864-023-09869-2</a> <a href="https://link.springer.com/article/10.1186/s12864-023-09869-2">https://link.springer.com/article/10.1186/s12864-023-09869-2</a>
          </p>
        </li>
        <li>
          <p>
	    Chen, S., Francioli, L.C., Goodrich, J.K. et al. A genomic mutational constraint map
	    using variation in 76,156 human genomes. Nature 625, 92–100
	    (2024). <a href="https://doi.org/10.1038/s41586-023-06045-0">https://doi.org/10.1038/s41586-023-06045-0</a> <a href="https://www.nature.com/articles/s41586-023-06045-0">https://www.nature.com/articles/s41586-023-06045-0</a>
          </p>
        </li>
        <li>
          <p>
	    Mosca, M.J., Cho, H. Reconstruction of private genomes through reference-based genotype
	    imputation. Genome Biol 24, 271
	    (2023). <a href="https://doi.org/10.1186/s13059-023-03105-6">https://doi.org/10.1186/s13059-023-03105-6</a> <a href="https://link.springer.com/article/10.1186/s13059-023-03105-6">https://link.springer.com/article/10.1186/s13059-023-03105-6</a>
          </p>
        </li>
        <li>
          <p>
	    Stöberl, N., Donaldson, J., Binda, C.S. et al. Mutant huntingtin confers cell-autonomous
	    phenotypes on Huntington’s disease iPSC-derived microglia. Sci Rep 13, 20477
	    (2023). <a href="https://doi.org/10.1038/s41598-023-46852-z">https://doi.org/10.1038/s41598-023-46852-z</a> <a href="https://www.nature.com/articles/s41598-023-46852-z">https://www.nature.com/articles/s41598-023-46852-z</a>
          </p>
        </li>
        <li>
          <p>
	    Tamman, A.J.F., Koller, D., Nagamatsu, S. et al. Psychosocial moderators of polygenic
	    risk scores of inflammatory biomarkers in relation to GrimAge. Neuropsychopharmacol. 49,
	    699–708
	    (2024). <a href="https://doi.org/10.1038/s41386-023-01747-5">https://doi.org/10.1038/s41386-023-01747-5</a> <a href="https://www.nature.com/articles/s41386-023-01747-5">https://www.nature.com/articles/s41386-023-01747-5</a>
          </p>
        </li>
        <li>
          <p>
	    Mignogna, G., Carey, C.E., Wedow, R. et al. Patterns of item nonresponse behaviour to
	    survey questionnaires are systematic and associated with genetic loci. Nat Hum Behav 7,
	    1371–1387
	    (2023). <a href="https://doi.org/10.1038/s41562-023-01632-7">https://doi.org/10.1038/s41562-023-01632-7</a> <a href="https://www.nature.com/articles/s41562-023-01632-7">https://www.nature.com/articles/s41562-023-01632-7</a>
          </p>
        </li>
        <li>
          <p>
	    Al-Jumaan, M., Chu, H., Alsulaiman, A. et al. Interplay of Mendelian and polygenic risk
	    factors in Arab breast cancer patients. Genome Med 15, 65
	    (2023). <a href="https://doi.org/10.1186/s13073-023-01220-4">https://doi.org/10.1186/s13073-023-01220-4</a> <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-023-01220-4">https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-023-01220-4</a>
          </p>
        </li>
        <li>
          <p>
	    Ilves N, Pajusalu S, Kahre T, et al. High Prevalence of Collagenopathies in Preterm- and
	    Term-Born Children With Periventricular Venous Hemorrhagic Infarction. Journal of Child
	    Neurology. 2023;38(6-7):373-388. doi:10.1177/08830738231186233. <a href="https://journals.sagepub.com/doi/full/10.1177/08830738231186233">https://journals.sagepub.com/doi/full/10.1177/08830738231186233</a>
          </p>
        </li>
        <li>
          <p>
	    Mignogna, G., Carey, C.E., Wedow, R. et al. Patterns of item nonresponse behaviour to
	    survey questionnaires are systematic and associated with genetic loci. Nat Hum Behav 7,
	    1371–1387
	    (2023). <a href="https://doi.org/10.1038/s41562-023-01632-7">https://doi.org/10.1038/s41562-023-01632-7</a> <a href="https://www.nature.com/articles/s41562-023-01632-7">https://www.nature.com/articles/s41562-023-01632-7</a>
          </p>
        </li>
        <li>
          <p>
	    Josefine U Melchiorsen, Kimmie V Sørensen, Jette Bork-Jensen, Hüsün S Kizilkaya, Lærke S
	    Gasbjerg, Alexander S Hauser, Jørgen Rungby, Henrik T Sørensen, Allan Vaag, Jens S
	    Nielsen, Oluf Pedersen, Allan Linneberg, Bolette Hartmann, Anette P Gjesing, Jens J
	    Holst, Torben Hansen, Mette M Rosenkilde, Niels Grarup, Rare Heterozygous
	    Loss-of-Function Variants in the Human GLP-1 Receptor Are Not Associated With
	    Cardiometabolic Phenotypes, The Journal of Clinical Endocrinology & Metabolism, Volume
	    108, Issue 11, November 2023, Pages
	    2821–2833, <a href="https://doi.org/10.1210/clinem/dgad290">https://doi.org/10.1210/clinem/dgad290</a>. <a href="https://academic.oup.com/jcem/article/108/11/2821/7180819">https://academic.oup.com/jcem/article/108/11/2821/7180819</a>
          </p>
        </li>
        <li>
          <p>
	    Vukadinovic, Milos et al. Deep learning-enabled analysis of medical images identifies
	    cardiac sphericity as an early marker of cardiomyopathy and related outcomes. Med,
	    Volume 4, Issue 4, 252 - 262.e3. <a href="https://www.cell.com/med/fulltext/S2666-6340(23)00069-7">https://www.cell.com/med/fulltext/S2666-6340(23)00069-7</a>
          </p>
        </li>
        <li>
          <p>
	    Epi25 Collaborative; Chen S, Neale BM, Berkovic SF. Shared and distinct ultra-rare
	    genetic risk for diverse epilepsies: A whole-exome sequencing study of 54,423
	    individuals across multiple genetic ancestries. medRxiv [Preprint]. 2023 Feb
	    24:2023.02.22.23286310. doi: 10.1101/2023.02.22.23286310. PMID: 36865150; PMCID:
	    PMC9980234. <a href="https://pubmed.ncbi.nlm.nih.gov/36865150/">https://pubmed.ncbi.nlm.nih.gov/36865150/</a>
          </p>
        </li>
        <li>
          <p>
	    Kurki, M.I., Karjalainen, J., Palta, P. et al. FinnGen provides genetic insights from a
	    well-phenotyped isolated population. Nature 613, 508–518
	    (2023). <a href="https://doi.org/10.1038/s41586-022-05473-8">https://doi.org/10.1038/s41586-022-05473-8</a> <a href="https://www.nature.com/articles/s41586-022-05473-8">https://www.nature.com/articles/s41586-022-05473-8</a>
          </p>
        </li>
        <li>
          <p>
	    Mortensen, Ó., Thomsen, E., Lydersen, L.N. et al. FarGen: Elucidating the distribution
	    of coding variants in the isolated population of the Faroe Islands. Eur J Hum Genet 31,
	    329–337
	    (2023). <a href="https://doi.org/10.1038/s41431-022-01227-2">https://doi.org/10.1038/s41431-022-01227-2</a> <a href="https://www.nature.com/articles/s41431-022-01227-2">https://www.nature.com/articles/s41431-022-01227-2</a>
          </p>
        </li>
        <li>
          <p>
	    Steiner, H.E., Carrion, K.C., Giles, J.B., Lima, A.R., Yee, K., Sun, X., Cavallari,
	    L.H., Perera, M.A., Duconge, J. and Karnes, J.H. (2023), Local Ancestry-Informed
	    Candidate Pathway Analysis of Warfarin Stable Dose in Latino Populations. Clin Pharmacol
	    Ther, 113:
	    680-691. <a href="https://doi.org/10.1002/cpt.2787">https://doi.org/10.1002/cpt.2787</a> <a href="https://ascpt.onlinelibrary.wiley.com/doi/full/10.1002/cpt.2787">https://ascpt.onlinelibrary.wiley.com/doi/full/10.1002/cpt.2787</a>
          </p>
        </li>
      </ul>
      <h2 id="section">2022</h2>
      <ul>
        <li>
          <p>
	    Huang, J., Tao, Q., Ang, T.F.A. et al. The impact of increasing levels of blood
	    C-reactive protein on the inflammatory loci SPI1 and CD33 in Alzheimer’s disease. Transl
	    Psychiatry 12, 523
	    (2022). <a href="https://doi.org/10.1038/s41398-022-02281-6">https://doi.org/10.1038/s41398-022-02281-6</a> <a href="https://www.nature.com/articles/s41398-022-02281-6">https://www.nature.com/articles/s41398-022-02281-6</a>
          </p>
        </li>
        <li>
          <p>
            Wadon, M.E., Fenner, E., Kendall, K.M. et al. Clinical and genotypic analysis in
	    determining dystonia non-motor phenotypic heterogeneity: a UK Biobank study. J Neurol
	    269, 6436–6451 (2022). <a href="https://doi.org/10.1007/s00415-022-11307-4">https://doi.org/10.1007/s00415-022-11307-4</a> <a href="https://link.springer.com/article/10.1007/s00415-022-11307-4">https://link.springer.com/article/10.1007/s00415-022-11307-4</a>
          </p>
        </li>
        <li>
          <p>
	    Andi Madihah Manggabarani, Takuyu Hashiguchi, Masatsugu Hashiguchi, Atsushi Hayashi,
	    Masataka Kikuchi, Yusdar Mustamin, Masaru Bamba, Kunihiro Kodama, Takanari Tanabata,
	    Sachiko Isobe, Hidenori Tanaka, Ryo Akashi, Akihiro Nakaya, Shusei Sato, Construction of
	    prediction models for growth traits of soybean cultivars based on phenotyping in diverse
	    genotype and environment combinations, DNA Research, Volume 29, Issue 4, August 2022,
	    dsac024, <a href="https://doi.org/10.1093/dnares/dsac024">https://doi.org/10.1093/dnares/dsac024</a> <a href="https://academic.oup.com/dnaresearch/article/29/4/dsac024/6653298?login=false">https://academic.oup.com/dnaresearch/article/29/4/dsac024/6653298?login=false</a>
          </p>
        </li>
        <li>
          <p>
	    Chaffin, M., Papangeli, I., Simonson, B. et al. Single-nucleus profiling of human
	    dilated and hypertrophic cardiomyopathy. Nature 608, 174–180
	    (2022). <a href="https://doi.org/10.1038/s41586-022-04817-8">https://doi.org/10.1038/s41586-022-04817-8</a> <a href="https://www.nature.com/articles/s41586-022-04817-8">https://www.nature.com/articles/s41586-022-04817-8</a>
          </p>
        </li>
        <li>
          <p>
	    Lee, J., Lee, J., Jeon, S. et al. A database of 5305 healthy Korean individuals reveals
	    genetic and clinical implications for an East Asian population. Exp Mol Med 54,
	    1862–1871
	    (2022). <a href="https://doi.org/10.1038/s12276-022-00871-4">https://doi.org/10.1038/s12276-022-00871-4</a> <a href="https://www.nature.com/articles/s12276-022-00871-4">https://www.nature.com/articles/s12276-022-00871-4</a>
          </p>
        </li>
        <li>
          <p>
	    Akingbuwa, W.A., Hammerschlag, A.R., Bartels, M. et al. Ultra-rare and common genetic
	    variant analysis converge to implicate negative selection and neuronal processes in the
	    aetiology of schizophrenia. Mol Psychiatry 27, 3699–3707
	    (2022). <a href="https://doi.org/10.1038/s41380-022-01621-8">https://doi.org/10.1038/s41380-022-01621-8</a> <a href="https://www.nature.com/articles/s41380-022-01621-8">https://www.nature.com/articles/s41380-022-01621-8</a>
          </p>
        </li>
        <li>
          <p>
	    Mitja, K.I., et al. FinnGen: Unique genetic insights from combining isolated population
	    and national health register data. medRxiv 2022.03.03.22271360;
	    doi: <a href="https://doi.org/10.1101/2022.03.03.22271360">https://doi.org/10.1101/2022.03.03.22271360</a>. <a href="https://www.medrxiv.org/content/10.1101/2022.03.03.22271360v1">https://www.medrxiv.org/content/10.1101/2022.03.03.22271360v1</a>
          </p>
        </li>
        <li>
          <p>
	    Akingbuwa, O. A. (2022). Polygenic analyses of childhood and adult psychopathology, and
	    their overlap. [PhD- Thesis - Research and graduation internal, Vrije Universiteit
	    Amsterdam]. <a href="https://research.vu.nl/ws/portalfiles/portal/149553301/O+A++Akingbuwa+-+thesis.pdf">https://research.vu.nl/ws/portalfiles/portal/149553301/O+A++Akingbuwa+-+thesis.pdf</a>
          </p>
        </li>
      </ul>
      <h2 id="section">2021</h2>
      <ul>
        <li>
          <p>
            Atkinson, E.G., et al. "Tractor uses local ancestry to enable the inclusion of admixed individuals in GWAS and to boost power", <em>Nature Genetics</em> (2021).
            <a href="https://doi.org/10.1038/s41588-020-00766-y" class="uri">https://doi.org/10.1038/s41588-020-00766-y</a>
            <a href="https://www.nature.com/articles/s41588-020-00766-y" class="uri">https://www.nature.com/articles/s41588-020-00766-y</a>
          </p>
        </li>
        <li><p>Maes, H.H. "Notes on Three Decades of Methodology Workshops", <em>Behavior Genetics</em> (2021). <a href="https://doi.org/10.1007/s10519-021-10049-9" class="uri">https://doi.org/10.1007/s10519-021-10049-9</a> <a href="https://link.springer.com/article/10.1007/s10519-021-10049-9" class="uri">https://link.springer.com/article/10.1007/s10519-021-10049-9</a></p></li>
        <li><p>Malanchini, M., et al. "Pathfinder: A gamified measure to integrate general cognitive ability into the biological, medical and behavioural sciences.", <em>bioRxiv</em> (2021). <a href="https://www.biorxiv.org/content/10.1101/2021.02.10.430571v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/2021.02.10.430571v1.abstract</a> <a href="https://www.biorxiv.org/content/10.1101/2021.02.10.430571v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/2021.02.10.430571v1.abstract</a></p></li>
      </ul>
      <h2 id="section">2020</h2>
      <ul>
        <li><p>Zekavat, S.M., et al. "Hematopoietic mosaic chromosomal alterations and risk for infection among 767,891 individuals without blood cancer", <em>medRxiv</em> (2020). <a href="https://doi.org/10.1101/2020.11.12.20230821" class="uri">https://doi.org/10.1101/2020.11.12.20230821</a> <a href="https://europepmc.org/article/ppr/ppr238896" class="uri">https://europepmc.org/article/ppr/ppr238896</a></p></li>
        <li><p>Kwong, A.K., et al. "Exome Sequencing in Paediatric Patients with Movement Disorders with Treatment Possibilities", <em>Research Square</em> (2020). <a href="https://doi.org/10.21203/rs.3.rs-101211/v1" class="uri">https://doi.org/10.21203/rs.3.rs-101211/v1</a> <a href="https://europepmc.org/article/ppr/ppr235428" class="uri">https://europepmc.org/article/ppr/ppr235428</a></p></li>
        <li><p>Krissaane, I, et al. “Scalability and cost-effectiveness analysis of whole genome-wide association studies on Google Cloud Platform and Amazon Web Services”, <em>Journal of the American Medical Informatics Association</em> (2020) ocaa068 <a href="https://doi.org/10.1093/jamia/ocaa068" class="uri">https://doi.org/10.1093/jamia/ocaa068</a> <a href="https://academic.oup.com/jamia/article/doi/10.1093/jamia/ocaa068/5876972" class="uri">https://academic.oup.com/jamia/article/doi/10.1093/jamia/ocaa068/5876972</a></p></li>
        <li><p>Karaca M, Atceken N, Karaca Ş, Civelek E, Şekerel BE, Polimanti R. “Phenotypic and Molecular Characterization of Risk Loci Associated With Asthma and Lung Function” <em>Allergy Asthma Immunol Res.</em> (2020) 12(5):806-820. <a href="https://doi.org/10.4168/aair.2020.12.5.806" class="uri">https://doi.org/10.4168/aair.2020.12.5.806</a> <a href="https://e-aair.org/DOIx.php?id=10.4168/aair.2020.12.5.806" class="uri">https://e-aair.org/DOIx.php?id=10.4168/aair.2020.12.5.806</a></p></li>
        <li><p>Muniz Carvalho, C., Wendt, F.R., Maihofer, A.X. et al. Dissecting the genetic association of C-reactive protein with PTSD, traumatic events, and social support. <em>Neuropsychopharmacol.</em> (2020). <a href="https://doi.org/10.1038/s41386-020-0655-6" class="uri">https://doi.org/10.1038/s41386-020-0655-6</a> <a href="https://www.nature.com/articles/s41386-020-0655-6#citeas" class="uri">https://www.nature.com/articles/s41386-020-0655-6#citeas</a></p></li>
      </ul>
      <h2 id="section-1">2019</h2>
      <ul>
        <li><p>Farhan, Sali MK, et al. “Exome sequencing in amyotrophic lateral sclerosis implicates a novel gene, DNAJC7, encoding a heat-shock protein” <em>Nature Neuroscience</em> (2019): 307835. <a href="https://www.nature.com/articles/s41593-019-0530-0" class="uri">https://www.nature.com/articles/s41593-019-0530-0</a></p></li>
        <li><p>Gay, Nicole R. et al. “Impact of admixture and ancestry on eQTL analysis and GWAS colocalization in GTEx” <em>bioRxiv</em> (2019) 836825; <a href="https://www.biorxiv.org/content/10.1101/836825v1" class="uri">https://www.biorxiv.org/content/10.1101/836825v1</a></p></li>
        <li><p>Sakaue, Saori et al. “Trans-biobank analysis with 676,000 individuals elucidates the association of polygenic risk scores of complex traits with human lifespan” <em>bioRxiv</em> (2019): 856351 <a href="https://www.biorxiv.org/content/10.1101/856351v1" class="uri">https://www.biorxiv.org/content/10.1101/856351v1</a></p></li>
        <li><p>Polimanti, Renato et al. “Leveraging genome-wide data to investigate differences between opioid use vs. opioid dependence in 41,176 individuals from the Psychiatric Genomics Consortium” <em>bioRxiv</em> (2019): 765065 <a href="https://www.biorxiv.org/content/10.1101/765065v1" class="uri">https://www.biorxiv.org/content/10.1101/765065v1</a></p></li>
        <li><p>Lescai, Francesco et al. “Meta-analysis of Scandinavian Schizophrenia Exomes” <em>bioRxiv</em> (2019): 836957; <a href="https://www.biorxiv.org/content/10.1101/836957v2" class="uri">https://www.biorxiv.org/content/10.1101/836957v2</a></p></li>
        <li><p>Bolze, Alexandre, et al. “Selective constraints and pathogenicity of mitochondrial DNA variants inferred from a novel database of 196,554 unrelated individuals” <em>bioRxiv</em> (2019): 798264;<a href="https://www.biorxiv.org/content/10.1101/798264v1" class="uri">https://www.biorxiv.org/content/10.1101/798264v1</a></p></li>
        <li><p>De Lillo, A., De Angelis, F., Di Girolamo, M. et al. “Phenome-wide association study of TTR and RBP4 genes in 361,194 individuals reveals novel insights in the genetics of hereditary and wildtype transthyretin amyloidoses.” <em>Hum Genet</em> 138, 1331–1340 (2019). <a href="https://www.ncbi.nlm.nih.gov/pubmed/31659433" class="uri">https://www.ncbi.nlm.nih.gov/pubmed/31659433</a></p></li>
        <li><p>Pividori, Milton, et al. “Shared and distinct genetic risk factors for childhood-onset and adult-onset asthma: genome-wide and transcriptome-wide studies.” <em>The Lancet Respiratory Medicine</em> 7.6 (2019): 509-522. <a href="https://www.biorxiv.org/content/10.1101/427427v2" class="uri">https://www.biorxiv.org/content/10.1101/427427v2</a></p></li>
        <li><p>Werling, Donna, et al. “Whole-genome and RNA sequencing reveal variation and transcriptomic coordination in the developing human prefrontal cortex.” <em>bioRxiv</em> (2019): 538421. <a href="https://www.biorxiv.org/content/10.1101/585430v1" class="uri">https://www.biorxiv.org/content/10.1101/585430v1</a></p></li>
        <li><p>Satterstrom, Kyle F., et al. “Large-scale exome sequencing study implicates both developmental and functional changes in the neurobiology of autism.” <em>bioRxiv</em> (2019): 538421. <a href="https://www.biorxiv.org/content/10.1101/484113v3" class="uri">https://www.biorxiv.org/content/10.1101/484113v3</a></p></li>
        <li><p>Huang, Qin, et al. “Delivering genes across the blood-brain barrier: LY6A, a novel cellular receptor for AAV-PHP. B capsids.” <em>bioRxiv</em> (2019): 538421. <a href="https://www.biorxiv.org/content/10.1101/538421v1" class="uri">https://www.biorxiv.org/content/10.1101/538421v1</a></p></li>
        <li><p>Kurki, Mitja I., et al. “Contribution of rare and common variants to intellectual disability in a sub-isolate of Northern Finland.” <em>Nature Communications</em> 10.1 (2019): 410. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/30679432/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/30679432/</a></p></li>
        <li><p>Martin, Alicia R., et al. “Current clinical use of polygenic scores will risk exacerbating health disparities.” <em>bioRxiv</em> (2019): 441261. <a href="https://www.biorxiv.org/content/10.1101/441261v3" class="uri">https://www.biorxiv.org/content/10.1101/441261v3</a></p></li>
        <li><p>Collaborative, Epi25, et al. “Ultra-rare genetic variation in the epilepsies: a whole-exome sequencing study of 17,606 individuals.” <em>American Journal of Human Genetics</em> (2019): <a href="https://www.cell.com/ajhg/fulltext/S0002-9297(19)30207-1" class="uri">https://www.cell.com/ajhg/fulltext/S0002-9297(19)30207-1</a></p></li>
        <li><p>Karczewski, Konrad J., et al. “The mutational constraint spectrum quantified from variation in 141,456 humans.” <em>bioRxiv</em> (2019): 531210. <a href="https://www.biorxiv.org/content/10.1101/531210v4" class="uri">https://www.biorxiv.org/content/10.1101/531210v4</a></p></li>
        <li><p>Whiffin, Nicola, et al. “Human loss-of-function variants suggest that partial LRRK2 inhibition is a safe therapeutic strategy for Parkinsons disease.” <em>bioRxiv</em>() (2019): 561472. <a href="https://www.biorxiv.org/content/10.1101/561472v1" class="uri">https://www.biorxiv.org/content/10.1101/561472v1</a></p></li>
        <li><p>Cummings, Beryl B., et al. “Transcript expression-aware annotation improves rare variant discovery and interpretation.” <em>bioRxiv</em> (2019): 554444. <a href="https://www.biorxiv.org/content/10.1101/554444v1" class="uri">https://www.biorxiv.org/content/10.1101/554444v1</a></p></li>
        <li><p>Wang, Qingbo, et al. “Landscape of multi-nucleotide variants in 125,748 human exomes and 15,708 genomes.” <em>bioRxiv</em> (2019): 573378. <a href="https://www.biorxiv.org/content/10.1101/573378v2" class="uri">https://www.biorxiv.org/content/10.1101/573378v2</a></p></li>
        <li><p>Minikel, Eric Vallabh, et al. “Evaluating potential drug targets through human loss-of-function genetic variation.” <em>bioRxiv</em> (2019): 530881. <a href="https://www.biorxiv.org/content/10.1101/530881v2" class="uri">https://www.biorxiv.org/content/10.1101/530881v2</a></p></li>
        <li><p>Collins, Ryan L., et al. “An open resource of structural variation for medical and population genetics.” <em>bioRxiv</em> (2019): 578674. <a href="https://www.biorxiv.org/content/10.1101/578674v1" class="uri">https://www.biorxiv.org/content/10.1101/578674v1</a></p></li>
        <li><p>Whiffin, Nicola, et al. “Characterising the loss-of-function impact of 5’untranslated region variants in whole genome sequence data from 15,708 individuals.” <em>bioRxiv</em> (2019): 543504. <a href="https://www.biorxiv.org/content/10.1101/543504v1" class="uri">https://www.biorxiv.org/content/10.1101/543504v1</a></p></li>
        <li><p>Lacaze, Paul, et al. “The Medical Genome Reference Bank: a whole-genome data resource of 4000 healthy elderly individuals. Rationale and cohort design.” <em>European Journal of Human Genetics</em> 27.2 (2019): 308. <a href="https://www.nature.com/articles/s41431-018-0279-z" class="uri">https://www.nature.com/articles/s41431-018-0279-z</a></p></li>
        <li><p>Cirulli, Elizabeth T., et al. “Genome-wide rare variant analysis for thousands of phenotypes in 54,000 exomes.” <em>bioRxiv</em> (2019): 692368. <a href="https://www.biorxiv.org/content/10.1101/692368v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/692368v1.abstract</a></p></li>
        <li><p>Kerminen, Sini, et al. “Geographic Variation and Bias in the Polygenic Scores of Complex Diseases and Traits in Finland.” <em>American Journal of Human Genetics</em> (2019). <a href="https://www.biorxiv.org/content/10.1101/485441v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/485441v1.abstract</a></p></li>
        <li><p>Jiang, Fan, Kyle Ferriter, and Claris Castillo. “PIVOT: Cost-Aware Scheduling of Data-Intensive Applications in a Cloud-Agnostic System.” <a href="https://renci.org/wp-content/uploads/2019/02/Cloud_19.pdf" class="uri">https://renci.org/wp-content/uploads/2019/02/Cloud_19.pdf</a></p></li>
        <li><p>Pividori, Milton, et al. “Shared and distinct genetic risk factors for childhood-onset and adult-onset asthma: genome-wide and transcriptome-wide studies.” <em>Lancet Respiratory Medicine</em> 7.6 (2019): 509-522. <a href="https://www.biorxiv.org/content/10.1101/427427v2" class="uri">https://www.biorxiv.org/content/10.1101/427427v2</a></p></li>
        <li><p>Cox, Samantha L., et al. “Genetic contributions to variation in human stature in prehistoric Europe.” <em>bioRxiv</em> (2019): 690545. <a href="https://www.biorxiv.org/content/10.1101/690545v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/690545v1.abstract</a></p></li>
        <li><p>Abrar, Faheem. A Modular Parallel Pipeline Architecture for GWAS Applications in a Cluster Environment. Diss. University of Saskatchewan, 2019. <a href="https://harvest.usask.ca/handle/10388/12087" class="uri">https://harvest.usask.ca/handle/10388/12087</a></p></li>
        <li><p>Khera, Amit V., et al. “Whole-genome sequencing to characterize monogenic and polygenic contributions in patients hospitalized with early-onset myocardial infarction.” <em>Circulation</em> 139.13 (2019): 1593-1602. <a href="https://www.ahajournals.org/doi/10.1161/CIRCULATIONAHA.118.035658" class="uri">https://www.ahajournals.org/doi/10.1161/CIRCULATIONAHA.118.035658</a></p></li>
      </ul>
      <h2 id="section-2">2018</h2>
      <ul>
        <li><p>An, Joon-Yong, et al. “Genome-wide de novo risk score implicates promoter variation in autism spectrum disorder.” <em>Science</em> (2018): 1. <a href="https://science.sciencemag.org/content/362/6420/eaat6576.full" class="uri">https://science.sciencemag.org/content/362/6420/eaat6576.full</a></p></li>
        <li><p>Molnos, Sophie Claudia. Metabolites: implications in type 2 diabetes and the effect of epigenome-wide interaction with genetic variation. Diss. Technische Universität München, 2018. <a href="https://mediatum.ub.tum.de/1372795f" class="uri">https://mediatum.ub.tum.de/1372795f</a></p></li>
        <li><p>Bis, Joshua C., et al. “Whole exome sequencing study identifies novel rare and common Alzheimer’s-associated variants involved in immune response and transcriptional regulation.” <em>Molecular Psychiatry</em> (2018): 1. <a href="https://www.nature.com/articles/s41380-018-0112-7" class="uri">https://www.nature.com/articles/s41380-018-0112-7</a></p></li>
        <li><p>Gormley, Padhraig, et al. “Common variant burden contributes to the familial aggregation of migraine in 1,589 families.” <em>Neuron</em> 98.4 (2018): 743-753. <a href="https://www.ncbi.nlm.nih.gov/pubmed/30189203" class="uri">https://www.ncbi.nlm.nih.gov/pubmed/30189203</a></p></li>
        <li><p>Rivas, Manuel A., et al. “Insights into the genetic epidemiology of Crohn’s and rare diseases in the Ashkenazi Jewish population.” <em>PLoS Genetics</em> 14.5 (2018): e1007329. <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007329" class="uri">https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007329</a></p></li>
        <li><p>Satterstrom, F. Kyle, et al. “ASD and ADHD have a similar burden of rare protein-truncating variants.” <em>bioRxiv</em> (2018): 277707. <a href="https://www.biorxiv.org/content/10.1101/277707v1" class="uri">https://www.biorxiv.org/content/10.1101/277707v1</a></p></li>
        <li><p>Zekavat, Seyedeh M., et al. “Deep coverage whole genome sequences and plasma lipoprotein (a) in individuals of European and African ancestries.” <em>Nature Communications</em> 9.1 (2018): 2606. <a href="https://www.nature.com/articles/s41467-018-04668-w" class="uri">https://www.nature.com/articles/s41467-018-04668-w</a></p></li>
        <li><p>Natarajan, Pradeep, et al. “Deep-coverage whole genome sequences and blood lipids among 16,324 individuals.” <em>Nature Communications</em> 9.1 (2018): 3391. <a href="https://www.nature.com/articles/s41467-018-04668-w" class="uri">https://www.nature.com/articles/s41467-018-04668-w</a></p></li>
        <li><p>Ganna, Andrea, et al. “Quantifying the impact of rare and ultra-rare coding variation across the phenotypic spectrum.” <em>American Journal of Human Genetics</em> 102.6 (2018): 1204-1211. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5992130/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5992130/</a></p></li>
        <li><p>Khera, Amit V., et al. “Genome-wide polygenic scores for common diseases identify individuals with risk equivalent to monogenic mutations.” <em>Nature Genetics</em> 50.9 (2018): 1219. <a href="https://www.nature.com/articles/s41588-018-0183-z?_ga=2.263293700.980063710.1543017600-1151073636.1543017600" class="uri">https://www.nature.com/articles/s41588-018-0183-z?_ga=2.263293700.980063710.1543017600-1151073636.1543017600</a></p></li>
        <li><p>Roselli, Carolina, et al. “Multi-ethnic genome-wide association study for atrial fibrillation.” <em>Nature Genetics</em> 50.9 (2018): 1225. <a href="https://www.nature.com/articles/s41588-018-0133-9" class="uri">https://www.nature.com/articles/s41588-018-0133-9</a></p></li>
        <li><p>Arachchi, Harindra, et al. “matchbox: An open‐source tool for patient matching via the Matchmaker Exchange.” <em>Human Mutation</em> 39.12 (2018): 1827-1834. <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/humu.23655" class="uri">https://onlinelibrary.wiley.com/doi/abs/10.1002/humu.23655</a></p></li>
        <li><p>Laisk, Triin, et al. “GWAS meta-analysis highlights the hypothalamic-pituitary-gonadal axis (HPG axis) in the genetic regulation of menstrual cycle length.” <em>bioRxiv</em> (2018): 333708. <a href="https://www.biorxiv.org/content/10.1101/333708v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/333708v1.abstract</a></p></li>
        <li><p>Rees, Elliott, et al. “Association between schizophrenia and both loss of function and missense mutations in paralog conserved sites of voltage-gated sodium channels.” <em>bioRxiv</em> (2018): 246850. <a href="https://www.biorxiv.org/content/10.1101/246850v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/246850v1.abstract</a></p></li>
        <li><p>Haas, Mary E., et al. “Genetic association of albuminuria with cardiometabolic disease and blood pressure.” <em>American Journal of Human Genetics</em> 103.4 (2018): 461-473. <a href="https://www.cell.com/ajhg/pdf/S0002-9297(18)30270-2.pdf" class="uri">https://www.cell.com/ajhg/pdf/S0002-9297(18)30270-2.pdf</a></p></li>
        <li><p>Abel, Haley J., et al. “Mapping and characterization of structural variation in 17,795 deeply sequenced human genomes.” <em>bioRxiv</em> (2018): 508515. <a href="https://www.biorxiv.org/content/10.1101/508515v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/508515v1.abstract</a></p></li>
        <li><p>Lane, Jacqueline M., et al. “Biological and clinical insights from genetics of insomnia symptoms.” <em>bioRxiv</em> (2018): 257956. <a href="https://www.biorxiv.org/content/10.1101/257956v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/257956v1.abstract</a></p></li>
        <li><p>Pividori, Milton, et al. “Shared and distinct genetic risk factors for childhood onset and adult onset asthma.” <em>bioRxiv</em> (2018): 427427. <a href="https://www.biorxiv.org/content/10.1101/427427v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/427427v1.abstract</a></p></li>
      </ul>
      <h2 id="section-3">2017</h2>
      <ul>
        <li>Lessard, Samuel, et al. “Human genetic variation alters CRISPR-Cas9 on-and off-targeting specificity at therapeutically implicated loci.” <em>Proceedings of the National Academy of Sciences</em> 114.52 (2017): E11257-E11266. <a href="https://www.pnas.org/content/114/52/E11257.long" class="uri">https://www.pnas.org/content/114/52/E11257.long</a>
        </li>
      </ul>
      <h2 id="section-4">2016</h2>
      <ul>
        <li>Ganna, Andrea, et al. “Ultra-rare disruptive and damaging mutations influence educational attainment in the general population.” <em>Nature Neuroscience</em> 19.12 (2016): 1563. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5127781/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5127781/</a>
        </li>
      </ul>
      <p><em>Footnote</em> In addition to software development, the Hail team engages in theoretical, algorithmic, and empirical research inspired by scientific collaboration. Examples include <a href="https://github.com/danielkunin/Regularized-Linear-Autoencoders">Loss landscapes of regularized linear autoencoders</a>, <a href="https://github.com/jbloom22/DASH">Secure multi-party linear regression at plaintext speed</a>, and <a href="https://www.nature.com/articles/s41592-018-0054-7">A synthetic-diploid benchmark for accurate variant-calling evaluation</a>.</p>
    </div>
  </div>
</div>
{% endblock %}
