package org.broadinstitute.hail.vcf

import org.broadinstitute.hail.variant._
import scala.collection.JavaConverters._

class BufferedLineIterator(bit: BufferedIterator[String]) extends htsjdk.tribble.readers.LineIterator {
  override def peek(): String = bit.head

  override def hasNext: Boolean = bit.hasNext

  override def next(): String = bit.next()

  override def remove() { throw new UnsupportedOperationException }
}

class HtsjdkRecordReader(codec: htsjdk.variant.vcf.VCFCodec)
  extends AbstractRecordReader {
  override def readRecord(line: String): Option[(Variant, Iterator[Genotype])] = {
    val vc = codec.decode(line)
    if (vc.isBiallelic) {
      val variant = Variant(vc.getContig, vc.getStart, vc.getReference.getBaseString,
        vc.getAlternateAllele(0).getBaseString)
      val b = new GenotypeStreamBuilder(variant)
      Some((variant,
        for (g <- vc.getGenotypes.iterator.asScala) yield {
          val gt = if (g.isNoCall)
            -1
          else if (g.isHomRef)
            0
          else if (g.isHet)
            1
          else {
            assert(g.isHomVar)
            2
          }

          val ad = if (g.hasAD) {
            val gad = g.getAD
            (gad(0), gad(1))
          } else
            (0, 0)

          val dp = if (g.hasDP)
            g.getDP
          else
            0

          val pl = if (g.isCalled) {
            if (g.hasPL) {
              val gpl = g.getPL
              (gpl(0), gpl(1), gpl(2))
            } else
              (0, 0, 0) // FIXME
          } else
            null

          Genotype(gt, ad, dp, pl)
        }))
    } else
      None
  }
}

object HtsjdkRecordReaderBuilder extends AbstractRecordReaderBuilder {
  def result(headerLines: Array[String]): HtsjdkRecordReader = {
    val codec = new htsjdk.variant.vcf.VCFCodec()
    codec.readHeader(new BufferedLineIterator(headerLines.iterator.buffered))
    new HtsjdkRecordReader(codec)
  }
}
